### ISSUE TITLE
Rule to detect and suggest multiple overloads instead of nullable default parameters.

### ISSUE THREAD

**Issue by mjr4077au** [meziantou/Meziantou.Analyzer#1299]
**Is your feature request related to a problem? Please describe.**

In a class within a project of mine, I had a static factory method of the below:
```C#
internal static MWV Get(SmbiosInformation? smbiosInfo = null)
{
    return new(smbiosInfo ?? SmbiosInformation.Get());
}
```
I had an issue in that the value I was passing to `smbiosInfo` wasn't initialised via a static ctor in time and the compiler treated this as the equivalent of ending through null, therefore ultimately the call to `SmbiosInformation.Get()` was made twice. 

I adopted a pattern like the below and the compiler was able to give me CS8604 to let me know that I had a possible null reference argument for this parameter.
```C#
internal static MWV Get()
{
    return new(SmbiosInformation.Get());
}

internal static MWV Get(SmbiosInformation smbiosInfo)
{
    return new(smbiosInfo);
}
```

**Describe the solution you'd like**

I think such a situation would make for a great analyser rule, to favour having multiple overloads with strongly typed and non-nullable parameters over a nullable default. This was a silent bug that while didn't cause any runtime issues, did lead to behaviour that was not immediately apparent.

**Describe alternatives you've considered**

Manually auditing my codebase and making the appropriate corrections. I can do the right thing, but an analyser rule would ensure others don't make the same mistakes.

**Comment by meziantou** [comment]
I'm not sure to understand the rule you want. Can you provide a complete example of the error you get and how the compiler helps with the 2 overloads.

Also, I don't see how this can be generalized. Maybe the example will help.

**Comment by mjr4077au** [comment]
Consider the below example class:
```C#
using System.IO;

namespace Devicie.TelemetryCollector
{
    internal static class TestClass
    {
        internal static readonly string TestPath;

        internal static readonly FileInfo TestFileInfo = GetFileInfoForPath(TestPath);

        [System.Diagnostics.CodeAnalysis.SuppressMessage("Minor Code Smell", "S1075:URIs should not be hardcoded", Justification = "Shhhh...")]
        static TestClass()
        {
            TestPath = "C:\\Temp\\ConfiguredFile.txt";
        }

        private static FileInfo GetFileInfoForPath(string? path = null)
        {
            return new(path ?? "C:\\Temp\\DefaultFile.txt");
        }
    }
}
```
If you change `private static FileInfo GetFileInfoForPath(string? path = null)` to `private static FileInfo GetFileInfoForPath(string path)`, the `TestFileInfo` line will have a CS8604 error indicating a possible null reference assignment for the `path` parameter. With two overloads, this doesn't occur.

If we inspect the compiled source using ILSpy, we can see that this is indeed a genuine concern due to how the static ctor ultimately becomes ordered:

<img width="1149" height="447" alt="Image" src="https://github.com/user-attachments/assets/e29ca971-2790-4a5e-851e-dc03c23c0359" />

**Comment by meziantou** [comment]
If I understand correctly, what you are asking is a rule to validate static initialization ordering.
[MA0195](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0195.md) already handles a part of it. It could be improved to check static constructor in addition to the fields.

**Comment by mjr4077au** [comment]
I'm surprised that rule didn't fire as I'm using `<MeziantouAnalysisMode>all-errors</MeziantouAnalysisMode>` and the way the rule reads, it seems like it should have fired for the above class anyway, irrespective of any additional static ctor checks:
```C#
internal static readonly string TestPath;
internal static readonly FileInfo TestFileInfo = GetFileInfoForPath(TestPath);
```
Maybe it sees that `TestPath` is being initialised in a static ctor and doesn't report? The logic may need enhancing so it knows that static ctor initialisation occurs after direct static field initialisation.

Lastly as well while I've got you, I'm working hard with my employer to try and get you some ongoing sponsorship for the work you do. I'm hoping that can be sorted in the next two weeks.

**Comment by meziantou** [comment]
MA0195 only considers explicitly initialized fields. I will improve it to detect explicitly initialized fields in static constructors, so it should report your case.

> Lastly as well while I've got you, I'm working hard with my employer to try and get you some ongoing sponsorship for the work you do. I'm hoping that can be sorted in the next two weeks.

Thanks for your support ❤️
